docs/src/starknet/calldata-transformation.md

```diff
@@ -0,0 +1,206 @@
+  * signed integers (`i8`, `i16`, `i32`, `i64`, `i128`)
+  * unsigned integers (`u8`, `u16`, `u32`, `u64`, `u96`, `u128`, `u256`, `u384`, `u512`)
+  * `felt252` (numeric literals and so-called *shortstrings*)
+  * `ByteArray`
+  * `ContractAddress`
+  * `ClassHash`
+  * `StorageAddress`
+  * `EthAddress`
+  * `bytes31`
+  * `Array` - using `array![]` macro
+
+Numeric types (primitives and `felt252`) can be paseed with type suffix specified -\
+for example `--calldata 420_u64`.
+
+### More complex examples
+
+  1. `complex_fn` - different data types:
+
+  ```shell
+  $ sncast --account myuser \
+      call \
+      --url http://127.0.0.1:5050 \
+      --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+      --function complex_fn \
+      --calldata \
+          array![array![1, 2], array![3, 4, 5], array![6]], \
+          12, \
+          -128_i8, \
+          "Some string (a ByteArray)", \
+          ('a shortstring', 32_u32), \
+          true, \
+          0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff \
+      --block-id latest
+  ```
+
+  Note the commas separating the arguments.
+
+  2. `nested_struct_fn` - struct nesting:
+
+  ```shell
+  $ sncast --account myuser \
+      call \
+      --url http://127.0.0.1:5050 \
+      --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+      --function nested_struct_fn \
+      --calldata \
+          data_transformer_contract::NestedStructWithField { \
+              a: data_transformer_contract::SimpleStruct { a: 10 }, \
+              b: 12 \
+          } \
+      --block-id latest
+  ```
+
+## Caveats
+
+> 💡 **Info**
+> To force transformation of a single numeric literal, simply add a type suffix to it.
+
+> 💡 **Info**
+> To force transformation of a list of `felt252`, separate its elements with commas.
+
+Reasoning about calldata serialization logic is not always possible for Cast. **In particular, Cast doesn't verify serialized calldata against the ABI**.
+
+When given a calldata, Cast first tries to interpret it as serialized.
+When failed, in case of type or length mismatch, it will then try to interpret it as Cairo expressions, according to ABI.
+
+Thus, a literal that *could* be interpreted as a value of an appropriate Cairo type is not always going to be proper calldata. Let's see an example:
+
+Our `DataTransformerContract` exposes a method with signature:
+```rust
+fn u256_fn(self: @ContractState, a: u256) { ... }
+```
+
+We can write:
+```shell
+$ sncast --account myuser \
+  call \
+  --url http://127.0.0.1:5050 \
+  --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+  --function u256_fn \
+  --calldata 0x10 \
+  --block-id latest
+```
+
+Despite the call data being the proper literal of an `u256`, call fails.
+That's because Cast interpreted `0x10` as a hex-encoded `felt252`, not `u256`. A properly serialized `u256` should have two `felt252` components.
```

Comment:
I'd reword it and just say that cast will always treat the data as serialized unless explicitly told to use data transformer (with commas, suffixes etc). I appreciate the detailed explanation but average user will probably just skim through this page if anything.

Reply:
This section is meant to be in-depth and optional to read for a better understanding. I reworked some ohter sections to emphasize the transformation rules and make them more straightforward.